fix minor issues for ATen/ROCm

shintaro-iwasaki opened the pull request:
This patch fixes a few minor issues and introduces a few changes that are found in enabling ROCm compilation for ATen.

1. Minor type-related changes in `ATen/miopen/*`
This is to suppress compiler warnings.

~2. Hipify `ATen/naitve/miopen/*.cpp`
`ATen/native/miopen/*.cpp` includes "cuda/CUDAConfig.h", which should be `hip/HIPConfig.h` (though compilation succeeds without this change since currently `CUDAConfig.h` = `HIPConfig.h`).~

3. Update `gen.py` to include `hip/EmptyTensor.h` instead of `cuda/EmptyTensor.h` for HIP compilation
`RegisterCUDA.cpp` (for HIP) should include `hip/EmptyTensor.h` (though compilation succeeds without this change since currently `cuda/EmptyTensor.h` does not contain CUDA-specific logic).

4. Exclude the `USE_DIRECT_NVRTC` code when `USE_ROCM=0`.
Note that `USE_DIRECT_NVRTC` is always undefined for OSS compilation. It seems that this flag exists only for an internal purpose.

~5. Exclude `frexp()` for ROCm <= 3.10
a newer ROCm (i.e., officially supported ROCm versions) has `frexp()`, but an old ROCm (e.g., ROCm <= 3.10) doesn't. This preprocessor branch avoids compilation error for old ROCm (though such an old ROCm is not officially supported).~

6. Change an include path from `aten/src/ATen/` to `ATen/` in `SharedReduceOps.h`
This is, as far as I checked, the only place that includes `Aten` from `aten/src`. This change unifies the include format.

Reviewed By: malfet

Differential Revision: D33441758

malfet commented:
@jeffdaily can you please have a look over proposed changes?

jeffdaily commented on tools/amd_build/build_amd.py:
```diff
@@ -78,6 +78,7 @@
     "aten/src/ATen/cuda/*",
     "aten/src/ATen/native/cuda/*",
     "aten/src/ATen/native/cudnn/*",
+    "aten/src/ATen/native/miopen/*",
```
Why was this needed?

shintaro-iwasaki replied:
Internally, I added this to convert `#include <ATen/cuda/CUDAConfig.h>` into `#include <ATen/hip/HIPConfig.h>`, but if OSS does not require this (and obviously it is not needed), I have to understand the OSS build flow more. Please let me investigate it. Thank you for pointing this out, @jeffdaily !

shintaro-iwasaki replied:
I investigated the background a bit more. This is the current situation:
- "aten/src/ATen/native/miopen/*" is now not hipified (because this is not in `build_amd.py` above)
- `#include <ATen/cuda/CUDAConfig.h>` is directly exposed for HIP compilation (e.g., in `BatchNorm_miopen.cpp`)
-  Currently `ATen/cuda/CUDAConfig.h` = `ATen/hip/HIPConfig.h` and both are there , so it works without any problem for OSS build.

(Internally, we remove `cuda/*` for HIP compilation, which causes this issue here.)

---

As far as I built ATen for ROCm, this is the only code that includes unhipified `cuda/CUDAConfig.h` for HIP compilation. @jeffdaily Should I keep this change in this PR? Because compilation succeeds now, should I keep this part untouched?

(Note: changes in `gen.py` etc, which generate `RegisterCUDA.cpp`, are to remove an unhipified `ATen/cuda/EmptyTensor.h` path from HIP compilation, but `RegisterCUDA.cpp` is build successfully without such a change since now `ATen/cuda/EmptyTensor.h` can be compiled with HIP.)

jeffdaily replied:
Adding `aten/src/ATen/native/miopen/*` to the hipify script seems odd since it should be a directory containing fully-custom HIP code under source control.  Does it make sense to use `#include <ATen/hip/HIPConfig.h>` directly in any files under that `miopen/*` directory instead, rather than relying on hipify to convert the name of the CUDA header that really isn't expected to be there in the first place?

shintaro-iwasaki replied:
Thanks @jeffdaily.  I truly agree that this design looks nonsense, but it is nontrivial to use `#include <ATen/hip/HIPConfig.h>` instead of `#include <ATen/cuda/CUDAConfig.h>`. This `#include <ATen/cuda/CUDAConfig.h>` is also used for CUDA to know if MIOpen is enabled. I would like to know your idea.

---

Details: To create `miopen` function stubs for CUDA, `native/miopen/*` files (e.g., `BatchNorm_miopen.cpp`) use `AT_ROCM_ENABLED()`, which is defined in `cuda/CUDAConfig`.
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/miopen/BatchNorm_miopen.cpp#L9
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cuda/CUDAConfig.h.in#L11

This is the same design pattern as `cudnn`, for example. In this case, it looks natural since it includes `cuda/CUDAConfig.h` to get `AT_CUDNN_ENABLED()`.
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/BatchNorm.cpp#L6

jeffdaily replied:
Thank you for the details.  In that case, your PR changes should be okay.  Can you rebase and make sure the ROCm CI is passing?  Your PR picked up a bad commit that has since been backed out to fix.

jeffdaily replied:
When the file `aten/src/ATen/native/miopen/BatchNorm_miopen.cpp` is hipified, it will end up one directory deeper at `aten/src/ATen/native/miopen/hip/BatchNorm_miopen.cpp`.  Which file is compiled by CMake?  When I build your PR locally I get the following object file, which indicates to me that your hipified version of BatchNorm_miopen.cpp is not the source file that gets compiled.

`./build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/ATen/native/miopen/BatchNorm_miopen.cpp.o`

shintaro-iwasaki replied:
Thanks. Please let me update this PR to address the issue (and I'll make sure the ROCm CI passes).

shintaro-iwasaki replied:
I updated `CMakeLists.txt`. Now it correctly compiles `miopen/hip/*`.

From the ROCm CI:
```
21:06:12 [ 88%] Building CXX object caffe2/CMakeFiles/torch_hip.dir/__/aten/src/ATen/native/miopen/hip/BatchNorm_miopen.cpp.o
21:06:12 [ 88%] Building CXX object caffe2/CMakeFiles/torch_hip.dir/__/aten/src/ATen/native/miopen/hip/RNN_miopen.cpp.o
```

shintaro-iwasaki replied:
@jeffdaily I'd appreciate it if you could review the change.

jeffdaily replied:
Please revert.

shintaro-iwasaki replied:
Thanks. Reverted the changes.

jeffdaily commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -126,6 +126,7 @@ file(GLOB native_quantized_cuda_cpp "native/quantized/cuda/*.cpp")
 file(GLOB native_hip_hip "native/hip/*.hip")
 file(GLOB native_hip_cpp "native/hip/*.cpp")
 file(GLOB native_miopen_cpp "native/miopen/*.cpp")
+file(GLOB native_miopen_hip_cpp "native/miopen/hip/*.cpp")
```
Please revert.

jeffdaily commented on aten/src/ATen/CMakeLists.txt:
```diff
@@ -223,7 +224,8 @@ if(USE_ROCM)
   # TODO: Codegen separate files for HIP and use those (s/cuda_generated_sources/hip_generated_sources)
   set(all_hip_cpp ${native_sparse_hip_cpp} ${native_quantized_hip_cpp} ${hip_cpp} ${native_hip_cpp} ${cuda_generated_sources} ${ATen_HIP_SRCS})
-  set(all_hip_cpp ${native_miopen_cpp} ${native_cudnn_hip_cpp} ${miopen_cpp} ${all_hip_cpp})
+  set(all_hip_cpp ${native_miopen_hip_cpp} ${native_cudnn_hip_cpp} ${miopen_cpp}
```
Please revert.

jeffdaily reviewed (changes requested):
Please revert the hipification of the directory `aten/src/ATen/native/miopen`.  The more I thought about it, it doesn't make sense to hipify a directory that already only builds when configured for ROCm.  Any necessary changes should be committed to the files directly.

I think to more correct fix would be to address the TODO comment in the various files under miopen (e.g., aten/src/ATen/native/miopen/RNN_miopen.cpp).

```
// TODO: Remove the condition on AT_ROCM_ENABLED entirely,
// don't build this file as part of CPU build.
#include <ATen/cuda/CUDAConfig.h>

#if !AT_ROCM_ENABLED()
```

> don't build this file as part of CPU build.

Can we split the miopen files and conditionally compile the stubs versus the actual?

shintaro-iwasaki commented:
@jeffdaily Thanks for your review. I reverted the change you mentioned. Since it does not affect the current OSS build, I would like to address it in another commit (if needed).

I will surely check if all the CI tests including ROCm pass, but I'd appreciate it if you could review the other changes too.

jeffdaily reviewed (approved):
LGTM, pending passing CI.

It would be nice to agree on the minimum ROCm version we must support in the sources.  This PR adds the ifdef to protect against frexp.  We recently restored this functionality in #67226, intentionally without ifdef'ing based on rocm version because we assumed no one was using rocm versions < 4.x.

xw285cornell commented:
@jeffdaily we're almost there in 4.5.2 upgrade. We shall add a todo here to clean it up.

shintaro-iwasaki committed:
fix minor issues for ATen/ROCm (#71925)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/71925

This patch fixes a few minor issues and introduces a few changes that are found in enabling ROCm compilation for ATen.

1. Minor type-related changes in `ATen/miopen/*`
This is to suppress compiler warnings.

2. Hipify `ATen/naitve/miopen/*.cpp`
`ATen/native/miopen/*.cpp` includes "cuda/CUDAConfig.h", which should be `hip/HIPConfig.h` (though compilation succeeds without this change since currently `CUDAConfig.h` = `HIPConfig.h`).

3. Update `gen.py` to include `hip/EmptyTensor.h` instead of `cuda/EmptyTensor.h` for HIP compilation
`RegisterCUDA.cpp` (for HIP) should include `hip/EmptyTensor.h` (though compilation succeeds without this change since currently `cuda/EmptyTensor.h` does not contain CUDA-specific logic).

4. Exclude the `USE_DIRECT_NVRTC` code when `USE_ROCM=0`.
Note that `USE_DIRECT_NVRTC` is always undefined for OSS compilation. It seems that this flag exists only for an internal purpose.

5. Exclude `frexp()` for ROCm <= 3.10
a newer ROCm (i.e., officially supported ROCm versions) has `frexp()`, but an old ROCm (e.g., ROCm <= 3.10) doesn't. This preprocessor branch avoids compilation error for old ROCm (though such an old ROCm is not officially supported).

6. Change an include path from `aten/src/ATen/` to `ATen/` in `SharedReduceOps.h`
This is, as far as I checked, the only place that includes `Aten` from `aten/src`. This change unifies the include format.

Test Plan: CI (including GitHub CI for ROCm)

Reviewed By: xw285cornell

Differential Revision: D33441758

fbshipit-source-id: d559c6a81edcf5cf2560ddd211551999667946ee

shintaro-iwasaki commented:
Thanks. Please let me remove the "HIP <=3.10" `#ifdef` from this PR.
